ext/nnstreamer/tensor_converter/tensor_converter_python3.cc

```diff
@@ -0,0 +1,494 @@
+}
+
+/**
+ * @brief	PYConverterCore Destructor
+ * @return	Nothing
+ */
+PYConverterCore::~PYConverterCore ()
+{
+  if (core_obj)
+    Py_XDECREF (core_obj);
+  if (shape_cls)
+    Py_XDECREF (shape_cls);
+  PyErr_Clear ();
+
+  dlclose (handle);
+  g_mutex_clear (&py_mutex);
+}
+
+/**
+ * @brief parse the invoke result to feed output tensors
+ * @param[result] Python object retunred by invoke
+ * @param[config] config Structure for output tensors config
+ * @return 0 if no error, otherwise negative errno
+ */
+int
+PYConverterCore::parseTensorsInfo (PyObject *result, PyObject *tensor_type, GstTensorsInfo *info)
+{
+  if (PyList_Size (result) < 0)
+    return -1;
+
+  info->num_tensors = PyList_Size (result);
+  for (unsigned int i = 0; i < info->num_tensors; i++) {
+    /** don't own the reference */
+    PyObject *tensor_shape = PyList_GetItem (result, (Py_ssize_t)i);
+    if (nullptr == tensor_shape)
+      throw std::runtime_error ("parseTensorsInfo() has failed (1).");
+
+    PyObject *shape_dims = PyObject_CallMethod (tensor_shape, (char *)"getDims", NULL);
+    if (nullptr == shape_dims)
+      throw std::runtime_error ("parseTensorsInfo() has failed (2).");
+
+    /** convert numpy type to tensor type */
+    info->info[i].type = (_nns_tensor_type) PyLong_AsLong (PyList_GetItem (tensor_type, (Py_ssize_t)i));
+    for (int j = 0; j < PyList_Size (shape_dims); j++)
+      info->info[i].dimension[j]
+          = (uint32_t)PyLong_AsLong (PyList_GetItem (shape_dims, (Py_ssize_t)j));
+
+    info->info[i].name = g_strdup("");
+    Py_XDECREF (shape_dims);
+  }
+
+  return 0;
+}
+
+/**
+ * @brief	convert any media stream to tensor
+ * @param[in] input : The array of input tensors
+ * @param[out]  output : The array of output tensors
+ * @return 0 if OK. non-zero if error.
+ *        -1 if the script does not work properly.
+ *        -2 if the output properties are different with script.
+ */
+GstBuffer *
+PYConverterCore::convert (GstBuffer *in_buf, GstTensorsConfig *config)
+{
+  GstMemory *in_mem, *out_mem;
+  GstMapInfo in_info;
+  GstBuffer *out_buf = NULL;
+  PyObject *tensors_info = NULL, *output = NULL, *tensor_type = NULL, *pyValue = NULL;
+  gint rate_n, rate_d;
+  guint mem_size;
+  gpointer mem_data;
+  if (nullptr == in_buf)
+    throw std::invalid_argument ("Null pointers are given to PYConverterCore::convert().\n");
+
+  in_mem = gst_buffer_peek_memory (in_buf, 0);
+
+  if (!gst_memory_map (in_mem, &in_info, GST_MAP_READ)) {
+    ml_loge ("Cannot map input memory / tensor_converter::custom-script");
+    return NULL;
+  }
+
+  npy_intp input_dims[] = { (npy_intp) (gst_buffer_get_size (in_buf)) };
+
+  PyObject *param = PyList_New (0);
+  PyObject *input_array = PyArray_SimpleNewFromData (
+      1, input_dims, NPY_UINT8, in_info.data);
+  PyList_Append (param, input_array);
+
+  Py_LOCK ();
+  if (!PyObject_HasAttrString (core_obj, (char *)"convert")) {
+    Py_ERRMSG ("Cannot find 'invoke'");
+    goto done;
```

Comment:
Please check the pairs of `Py_LOCK` and `PY_LOCK` in this function.

Reply:
Right. Updated.